### ISSUE TITLE
Add string interpolation support

### ISSUE THREAD

**Issue by nathanielc** [influxdata/ifql#313]
From the spec

String literals are also interpolated for embedded expressions to be evaluated as strings.
Embedded expressions are enclosed in curly brackets "{}".
The expressions are evaluated in the scope containing the string literal.
The result of an expression is formatted as a string and replaces the string content between the brackets.
All types are formatted as strings according to their literal representation.
A function "printf" exists to allow more precise control over formatting of various types.
To include the literal curly brackets within a string they must be escaped.

Interpolation example:

    n = 42
    "the answer is {n}" // the answer is 42
    "the answer is not {n+1}" // the answer is not 43
    "openinng curly bracket \{" // openinng curly bracket {
    "closing curly bracket \}" // closing curly bracket }

**Comment by nathanielc** [comment]
@mark-rushakoff I am referencing your comment here https://github.com/influxdata/ifql/pull/303#issuecomment-388196397

> I think the spec (and the definition of string_lit in particular) needs to clarify that a backslash followed by any other character than what is specified, is an invalid string (a compile-time error)

 +1

Nested quotes should be supported. Any valid expression should be allowed within the curly brackets. 
As such these should be valid:

```
sayTwice = (s) => "{s}{s}"

laughter = "{sayTwice(s:"ha")}!"
```

This technically means that you could have nested string interpolations. While I can't think of a good use case here is a trivial example:

```
name = "bob"
age = 24
s = "hi, {"my name is {name} and I am {age}"}"
// This can obviously be simplified to
s = "hi my name is {name} and I am {age}"
```

I don't think its worth restricting nested interpolated strings because that means going out of our way to enforce such a rule. Maybe when we go to implement we will learn if its feasible to easily restrict and if it provides any value to do so.

> I don't like that plain strings can have newlines embedded. ...

I agree, my concern is that the precedent in other languages is that multi-line strings are also raw strings, without any escape sequences or interpolation. In IFQL the most common case for using string interpolation will be with a large multi-line string. Specifically we see this a lot in TICKscript, were a user defines a really long formatted html message for an alert. The message in these cases nearly always contains TICKscript's form of string interpolation. 

In other words we need to support string interpolation in multi-line strings. That doesn't mean we need to support all escape sequences etc in multi-line strings, but at that point the difference was so small that I didn't think the distinction was worth creating.

Thoughts?

**Comment by mark-rushakoff** [comment]
> I don't think its worth restricting nested interpolated strings 

Agreed. I was more interested in having the `laughter = "{sayTwice(s:"ha")}!"` example to make it clear that quotes inside an interpolation context are still valid; there isn't some kind of syntax error following a literal string `"{sayTwice(s:"`.

Turns out Ruby does support multiline strings:

```ruby
s = "123
456"
```

Python does not support multiline strings with one set of quotes, but it does support multiline strings with triple quotes:

```python
s = """123
456"""
```

I can live with embedded newlines as currently specified. It would be good to hear from other team members on this one.

**Comment by mark-rushakoff** [comment]
I'm having a little trouble following the spec through what constitutes an Expression, as it relates to whitespace and comments in the interpolated scope of a string. Is this equivalent to the one-line example?

```
laughter = "{
 // We are now in the string interpolation scope.

 sayTwice(s: "ha")

 // We are now leaving the string interpolation scope.
}!"
```

If this is valid, we should have a (less-contrived?) string interpolation example with whitespace and comments.

**Comment by nathanielc** [comment]
The Expressions haven't been fully defined yet in the spec. Mostly because I want to get it right and the current PEG grammar is not EBNF and the translation is not trivial. See #318 

Yes that is an equivalent expression.

**Comment by adamperlin** [comment]
Kamino closed and cloned this issue to [influxdata/platform](https://github.com/influxdata/platform/issues/251)
